test/integrationtests/voight_kampff/tools.py

```diff
@@ -15,15 +15,161 @@
 """Common tools to use when creating step files for behave tests."""
 
 from threading import Event
+from typing import Any, Callable, List, Tuple
 import time
 
 from mycroft.audio.utils import wait_while_speaking
 from mycroft.messagebus import Message
 
-
 DEFAULT_TIMEOUT = 10
 
 
+class VoightKampffEventMatcher:
+    """Matches a specified message type to messages emitted on the bus.
+
+    Attributes:
+        message_type: identifier of the message to search for on the bus
+        context: the Behave context from the test utilizing this class
+        match_event: mechanism for knowing when a match is found
+        error_message: message that can be used by the test to communicate
+            the reason for a failed match to the tester.
+    """
+    def __init__(self, message_type: str, context: Any):
+        self.message_type = message_type
+        self.context = context
+        self.match_event = Event()
+        self.match_found = False
+        self.error_message = ""
+
+    def match(self, timeout: int = None):
+        """Attempts to match the requested message type to emitted bus events.
+
+        Use a message bus event handler to capture any message emitted on the
+        bus that matches the message type specified by the caller.  Also
+        checks any messages emitted prior to the handler being defined to
+        protect against a race condition.
+
+        Args:
+            timeout: number of seconds to attempt matching before giving up
+        """
+        timeout = timeout or self.context.step_timeout
+        self.context.bus.on(self.message_type, self.handle_message)
+        self._check_historical_messages()
+        if not self.match_event.is_set():
+            self.match_event.wait(timeout=timeout)
+        self.context.bus.remove(self.message_type, self.handle_message)
+        self.match_found = self.match_event.is_set()
+        if not self.match_found:
+            self._build_error_message()
+
+    def _check_historical_messages(self):
+        """Searches messages emitted before the event handler was defined."""
+        for message in self.context.bus.get_messages(self.message_type):
+            self.handle_message(message)
+            if self.match_event.is_set():
+                break
+        self.context.bus.clear_messages()
+
+    def handle_message(self, message: Message):
+        """Applies matching criteria to the emitted event.
+
+        Args:
+            message: message emitted by bus with the requested message type
+        """
+        self.context.matched_message = message
+        self.match_event.set()
+
+    def _build_error_message(self):
+        """Builds a message that communicates the failure to the test."""
+        self.error_message = (
+            f"Expected message type {self.message_type} was not emitted."
+        )
```

Comment:
Might be an idea to include the `reset()` method from the CriteraWaiter as well. so an object isn't left in an unusable state after a matching (unless you know how to reset it)

Reply:
These are intended to be single use classes.  Instantiate it, do the matching and be done with it.  I think reusing the same instance of a class for multiple check could cause trouble.  I could improve the comments to indicate this, or I could make the classes private and write wrapper functions that use them as intended.

Reply:
:heart:

Reply:
Explicitly stating that it's single use only sounds like a good idea.